pkg/kvstore/store/watchstore.go

```diff
@@ -0,0 +1,242 @@
+// information is synchronized from, which is used to enrich the metrics.
+func RWSWithSourceClusterName(cluster string) RWSOpt {
+	return func(rws *restartableWatchStore) {
+		rws.source = cluster
+	}
+}
+
+type rwsEntry struct {
+	key   Key
+	stale bool
+}
+
+// restartableWatchStore implements the WatchStore interface, supporting
+// multiple executions of the Watch() operation (granted that the previous one
+// already terminated). This allows to transparently handle the case in which
+// we had to create a new etcd connection (for instance following a failure)
+// which refers to the same remote cluster.
+type restartableWatchStore struct {
+	source     string
+	keyCreator KeyCreator
+	observer   Observer
+
+	watching        atomic.Bool
+	synced          bool
+	onSyncCallbacks []func(ctx context.Context)
+
+	// Using a separate entries counter avoids the need for synchronizing the
+	// access to the state map, since the only concurrent reader is represented
+	// by the NumEntries() function.
+	state      map[string]*rwsEntry
+	numEntries atomic.Uint64
+
+	log           *logrus.Entry
+	entriesMetric prometheus.Gauge
+}
+
+// NewRestartableWatchStore returns a WatchStore instance which supports
+// restarting the watch operation multiple times, automatically handling
+// the emission of deletion events for all stale entries (if enabled). It
+// shall be restarted only once the previous Watch execution terminated.
+func NewRestartableWatchStore(keyCreator KeyCreator, observer Observer, opts ...RWSOpt) WatchStore {
+	rws := &restartableWatchStore{
+		source:     option.Config.ClusterName,
+		keyCreator: keyCreator,
+		observer:   observer,
+
+		state: make(map[string]*rwsEntry),
+
+		log:           log,
+		entriesMetric: metrics.NoOpGauge,
+	}
+
+	for _, opt := range opts {
+		opt(rws)
+	}
+
+	rws.log = rws.log.WithField(logfields.ClusterName, rws.source)
+	return rws
+}
+
+// Watch starts watching the specified kvstore prefix, blocking until the context is closed.
+// It might be executed multiple times, granted that the previous execution already terminated.
+func (rws *restartableWatchStore) Watch(ctx context.Context, backend WatchStoreBackend, prefix string) {
+	// Append a trailing "/" to the prefix, to make sure that we watch only
+	// sub-elements belonging to that prefix, and not to sibling prefixes
+	// (for instance in case the last part of the prefix is the cluster name,
+	// and one is the substring of another).
+	if !strings.HasSuffix(prefix, "/") {
+		prefix = prefix + "/"
+	}
+
+	rws.log = rws.log.WithField(logfields.Prefix, prefix)
+	syncedMetric := metrics.KVStoreInitialSyncCompleted.WithLabelValues(
+		kvstore.GetScopeFromKey(prefix), rws.source, "read")
+
+	rws.log.Info("Starting restartable watch store")
+	syncedMetric.Set(metrics.BoolToFloat64(false))
+
+	if rws.watching.Swap(true) {
+		rws.log.Panic("Cannot start the watch store while still running")
```

Comment:
Sorry, I didn't get what this means (the one in the `Drain` as well). It's "restartable", but can't start while still running?